### PR TITLE
INTMDB-421 cloud backup restore jobs

### PR DESCRIPTION
**Pull request by rdharmedran**
## Description
 Feature update: cloud-provider-snapshot-restore-jobs

Link to any related issue(s): 
https://jira.mongodb.org/browse/INTMDB-421
## Type of change:

- [x]  Bug fix (non-breaking change which fixes an issue)
- [x]  New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Required Checklist:

- [x]  I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

## Further comments
 -  Renamed the module cloud-provider-snapshot-restore-jobs  into cloud-backup-restore- jobs
 - Added input parameter validation
 - Read function returned already deleted restore job.So we have added condition.
- Delete function was not checking if restore job is already exist before delete it.We have added condition for the same.
- Added cfn-test configurations.

### CODE REVIEW

**Comment by Zuhairahmed** [comment]
@rdharmedran can we add GitHub checks to this PR (build + linting)?

**Comment by rdharmedran** [comment]
> @rdharmedran can we add GitHub checks to this PR (build + linting)?

Yes -I have added now and committed files

File: cfn-resources/cloud-backup-restore-jobs/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# MongoDB::Atlas::CloudBackUpRestoreJobs
+
+## Description
+This resource allows you to create, cancel, get one or list all cloud backup restore jobs.
```
Comment (Zuhairahmed) [inline cfn-resources/cloud-backup-restore-jobs/README.md:4]:
```suggestion
Returns, starts, and cancels Cloud Backup restore jobs.
```
Use Open API Doc deflations as much as possible: https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Cloud-Backup-Restore-Jobs

File: cfn-resources/cloud-backup-restore-jobs/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+Both CREATE & DELETE tests must pass.
+
+## Installation
+TAGS=logging make
+cfn submit --verbose --set-default
+
+## Cloudformation Examples
+
+Please use the [CFN Template](test/cloud-backup-restore-job.sample-template.yaml)
```
Comment (Zuhairahmed) [inline cfn-resources/cloud-backup-restore-jobs/README.md:36]:
```suggestion
Please see the [CFN Template](test/cloud-backup-restore-job.sample-template.yaml) for example resource
```

File: cfn-resources/cloud-backup-restore-jobs/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+## Integration Testing w/ AWS
+
+The [/quickstart-mongodb-atlas/scripts/launch-x-quickstart.sh](launch-x-quickstart.sh) script
```
Comment (Zuhairahmed) [inline cfn-resources/cloud-backup-restore-jobs/README.md:40]: Link does not work

**Review by Zuhairahmed — COMMENTED** [review]
Few doc comments below

File: cfn-resources/cloud-backup-restore-jobs/README.md (hunk 1 above)
Comment (rdharmedran) [reply cfn-resources/cloud-backup-restore-jobs/README.md]: This issue is fixed

File: cfn-resources/cloud-backup-restore-jobs/README.md (hunk 2 above)
Comment (rdharmedran) [reply cfn-resources/cloud-backup-restore-jobs/README.md]: This issue is fixed

File: cfn-resources/cloud-backup-restore-jobs/README.md (hunk 3 above)
Comment (rdharmedran) [reply cfn-resources/cloud-backup-restore-jobs/README.md]: This issue is fixed

File: cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+	projectID                        = "ProjectId"
+	snapshotID                       = "SnapshotId"
+	clusterName                      = "ClusterName"
+	id                               = "Id"
+	errorCreateMongoClient           = "Error - Create MongoDB Client- Details: %+v"
+	errorCreateCloudBackupRestoreJob = "Error - Create Cloud Backup Restore snapshot for Snapshot(%s)- Details: %+v"
+	errorReadCloudBackUpRestoreJob   = "Error - Read Restore Job with id(%s)"
+	automated                        = "automated"
+	download                         = "download"
+)
```
Comment (swattyT) [inline cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go:28]: Can we utilize "github.com/mongodb/mongodbatlas-cloudformation-resources/util/constants" for consts?

File: cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+		JobID:       jobID,
+	}
+	// Check if snapshot already exist
+	if !isRestoreJobExist(currentModel) {
+		_, _ = logger.Warnf("Delete - errors reading restore with id(%s)", jobID)
+		return handler.ProgressEvent{
+			OperationStatus:  handler.Failed,
+			Message:          "Resource Not Found",
+			HandlerErrorCode: cloudformation.HandlerErrorCodeNotFound}, nil
+	}
```
Comment (swattyT) [inline cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go:210]: Please correct me if I am wrong but the logic for Read and Delete until "Check if snapshot already exists" seems the same to me. If so, is it possible to extract in a separate function and reuse it?

Comment (rdharmedran) [reply cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go:210]: The input field validation is different for read and delete and also log statements

File: cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go (hunk 4 above)
Comment (rdharmedran) [reply cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go:28]: sure we will move this

Comment (rdharmedran) [reply cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go]: This is fixed.Please check now

File: cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go (hunk 5 above)
Comment (rdharmedran) [reply cfn-resources/cloud-backup-restore-jobs/cmd/resource/resource.go]: This is fixed.Please check now

**Review by govardhanpagidi — APPROVED** [review]

**Review by Zuhairahmed — APPROVED** [review]
LGTM!

**Review by swattyT — APPROVED** [review]
LGTM!

**Pull request merged**